components/script/dom/documentorshadowroot.rs

```diff
@@ -0,0 +1,323 @@
+        document_element: Option<DomRoot<Element>>,
+    ) -> Option<DomRoot<Element>> {
+        // TODO: Step 2.
+
+        match focused_element {
+            Some(element) => Some(element), // Step 3. and 4.
+            None => match body {
+                // Step 5.
+                Some(body) => Some(DomRoot::upcast(body)),
+                None => document_element,
+            },
+        }
+    }
+
+    /// Remove a stylesheet owned by `owner` from the list of document sheets.
+    #[allow(unrooted_must_root)] // Owner needs to be rooted already necessarily.
+    pub fn remove_stylesheet(
+        owner: &Element,
+        s: &Arc<Stylesheet>,
+        mut stylesheets: StylesheetSet<StyleSheetInDocument>,
+    ) {
+        let guard = s.shared_lock.read();
+
+        // FIXME(emilio): Would be nice to remove the clone, etc.
+        stylesheets.remove_stylesheet(
+            None,
+            StyleSheetInDocument {
+                sheet: s.clone(),
+                owner: Dom::from_ref(owner),
+            },
+            &guard,
+        );
+    }
+
+    /// Add a stylesheet owned by `owner` to the list of document sheets, in the
+    /// correct tree position.
+    #[allow(unrooted_must_root)] // Owner needs to be rooted already necessarily.
+    pub fn add_stylesheet(
+        owner: &Element,
+        mut stylesheets: StylesheetSet<StyleSheetInDocument>,
+        sheet: Arc<Stylesheet>,
+        insertion_point: Option<StyleSheetInDocument>,
+        style_shared_lock: &StyleSharedRwLock,
+    ) {
+        // FIXME(emilio): It'd be nice to unify more code between the elements
+        // that own stylesheets, but StylesheetOwner is more about loading
+        // them...
+        debug_assert!(
+            owner.as_stylesheet_owner().is_some() || owner.is::<HTMLMetaElement>(),
+            "Wat"
+        );
+
+        let sheet = StyleSheetInDocument {
+            sheet,
+            owner: Dom::from_ref(owner),
+        };
+
+        let guard = style_shared_lock.read();
+
+        match insertion_point {
+            Some(ip) => {
+                stylesheets.insert_stylesheet_before(None, sheet, ip, &guard);
+            },
+            None => {
+                stylesheets.append_stylesheet(None, sheet, &guard);
+            },
+        }
+    }
+
+    /// Remove any existing association between the provided id and any elements in this document.
+    pub fn unregister_named_element(
+        &self,
+        id_map: &DomRefCell<HashMap<Atom, Vec<Dom<Element>>>>,
+        to_unregister: &Element,
+        id: &Atom,
+    ) {
+        debug!(
+            "Removing named element {:p}: {:p} id={}",
+            self, to_unregister, id
+        );
+        // Limit the scope of the borrow because id_map might be borrowed again by
+        // GetElementById through the following sequence of calls
+        // reset_form_owner_for_listeners -> reset_form_owner -> GetElementById
+        {
```

Comment:
I don't think this is relevant now that this is on its own function, so I'd remove the comment and the scope.